### DISCUSSION TITLE
New Built-in Link Component

### DISCUSSION THREAD

**Discussion by ghost** [withastro/roadmap#766] [Proposal]
## Summary

Introduce a built-in `Link` component in Astro for enhanced, automated handling of different types of links, streamlining development, and improving site performance.

## Background & Motivation

Currently, handling different link types in Astro requires manual effort and increases the complexity of link management, which can lead to errors and inefficiencies. This component addresses these challenges by providing a unified, intelligent way to handle links, enhancing developer experience and site performance.

## Goals

- **Simplify Link Handling:** Automate the differentiation between internal and external links, reducing manual coding effort and potential errors.
- **Enhance Link Functionality:** Intelligently handle special link types (e.g., `mailto:`, `tel:`) with appropriate SEO and performance attributes.
- **Improve User Navigation:** Automatically identify and style active links, enhancing user experience.
- **Boost SEO and Performance:** Apply relevant attributes (`rel="external"`, `rel="prefetch"`) to links based on their type, improving SEO and site performance.
- **Support Developer Productivity:** Reduce the redundancy in link-related code across different components or pages, contributing to more efficient and maintainable codebases.

## Example

Link.astro
```astro
---
import type { HTMLAttributes } from "astro/types";

type Props = HTMLAttributes<"a">;

const { href: rawHref, style, ...attrs } = Astro.props as Props;
const href = (rawHref instanceof URL ? rawHref.toString() : rawHref) ?? "";
const site = Astro.site;
const { pathname } = Astro.url;

function isActiveLink(url: string): boolean {
  return url === pathname || url === "." || url === "./";
}

function isExternalLink(url: string): boolean {
  return URL.canParse(url);
}

function isInternalLink(url: string): boolean {
  return URL.canParse(url, site?.toString());
}

function isSpecialLink(url: string): boolean {
  return (
    url.startsWith("mailto:") ||
    url.startsWith("tel:") ||
    url.startsWith("sms:") ||
    url.startsWith("#")
  );
}

const isExternal = isExternalLink(href);
const isInternal = isInternalLink(href);
const isSpecial = isSpecialLink(href);
const isActive = isActiveLink(href);

let relAttribute: object = {};

if (isExternal && !isSpecial) {
  relAttribute = { rel: "external noopener noreferrer" };
} else if (isInternal && !isSpecial && !isActive) {
  relAttribute = { rel: "prefetch" };
} else {
  relAttribute = {};
}
---

<a href={href} class:list={[{ active: isActive }]} {...relAttribute} {...attrs}>
  <slot />
</a>

```

Index.astro
```astro
---
import Link from '../components/Link.astro';

<Link href="https://example.com" set:text="Example" />
<Link href="/about" set:text="About" />
<Link href="mailto:contact@example.com" set:text="email us" />
---
```

This example illustrates the ease of using the `Link` component to handle various types of links, showcasing its versatility and utility in an Astro project.

**Comment by brunocascio** [comment]
I think what will be great is the ability of passing href as an object as [nextjs supports it](https://nextjs.org/docs/pages/api-reference/components/link#href-required). This is specially useful when working with query params
For example:

```jsx
<Link 
  href={{ 
    pathname: '/path', 
    query: { foo: 'bar' } 
  }}
>
  My Link with Query Params
</Link>
```

**Comment by delucis** [comment]
Could be great to see this published to npm as a user package people could install and use!

**Comment by ghost** [comment]
Updated [Link component](https://gist.github.com/jacobdalamb/6951a5bc07c9937f2f3991614b3685bc)

**Comment by eikowagenknecht** [comment]
Is there a reason you removed the handling of `mailto:` and other special links in the update?
